### ISSUE TITLE
Dataset splitting Procedures

### ISSUE THREAD

**Issue by rusty1s** [vrtex-team/pytorch_geometric#41]
Add DeepSnap splitting functionality to PyTorch Geometric. We still need to discuss on the API, e.g., whether we want to provide that set of functionalities in the graph storage implementation or via `transforms` (as we currently do in PyTorch Geometric, see [here](https://pytorch-geometric.readthedocs.io/en/latest/modules/transforms.html#torch_geometric.transforms.AddTrainValTestMask)).

**Comment by RexYing** [comment]
transform = T.RandomNodeSplit(inductive=False, num_splits=3): similar to existing implementation AddTrainValTestMask
Argument: split_ratio e.g. (0.8, 0.1, 0.1); minimum_per_split
data = transform(data)


transform = T.RandomNodeSplit(inductive=True, num_splits=3):
train_data, val_data, test_data = transform(data)
Val_data includes nodes in train_data and additional validation nodes
Test_data includes nodes in train_data and val_data, and additional test nodes


transform = T.RandomEdgeSplit(num_splits=3):
Similar to inductive RandomNodeSplit


transform = T.RandomGraphSplit(num_splits=3):
Split datasets according to __inc__

**Comment by rusty1s** [comment]
I think this is great. Here are some further thoughts:
1. Do we really want to include `train_data` in validation/tests splits for node classification? As far as I know, this isn't commonly done in literature.
2. I'm not sure what splitting according to `__inc__` actually means. I think for graph classification, the current interface of running `train_dataset = dataset[idx]` is actually fine. We just need some helper functions to give to calculate `idx`.

**Comment by RexYing** [comment]
There are 2 inductive node classification settings I'm testing on:  

- new nodes are added to the existing graph, in which case I need to keep the train_data nodes which are in the same graph
- new nodes all belong to a new graph, in which case train_data will not be needed.
Do you think they make sense? Or we should prioritize one of the scenarios first?

2. Yeah I agree. I was actually looking at the data object which is already collated into a single edge_index. But it makes more sense to do it at dataset level before collation.

**Comment by rusty1s** [comment]
I see. So the inductive scenario in which training nodes are kept during evaluation relates to some temporal data in which validation/test nodes do not exist yet during training, right? I think we can support both based on a simple flag (such as `keep_train_data`). I think the difference between the two approaches is that in the former, you still need to keep track of validation/test nodes via an additional mask attribute.
